src/components/patterns/headingList.tsx

```diff
@@ -0,0 +1,88 @@
+import React, {useEffect, useState} from "react"
+import {twMerge} from "tailwind-merge"
+
+interface Heading {
+  text: string
+  id: string
+}
+
+const HeadingList = () => {
+  const [headings, setHeadings] = useState<Heading[]>([])
+  const [activeHeading, setActiveHeading] = useState<string>("")
+
+  useEffect(() => {
+    const wysiwygContainers: NodeListOf<Element> = document.querySelectorAll(".wysiwyg")
+
+    const allHeadings: Heading[] = []
+
+    wysiwygContainers.forEach((container, containerIndex) => {
+      const h2Elements: HTMLHeadingElement[] = Array.from(container.querySelectorAll("h2"))
+
+      h2Elements.forEach((heading, headingIndex) => {
+        let id = heading.getAttribute("id")
+        if (!id) {
+          id = `${heading.textContent?.trim().toLowerCase().replace(/\s+/g, "-")}-${containerIndex}-${headingIndex}`
+          heading.setAttribute("id", id)
+        }
+
+        allHeadings.push({text: heading.textContent || "", id})
+      })
+    })
+
+    setHeadings(allHeadings)
+
+    const observerCallback: IntersectionObserverCallback = entries => {
+      entries.forEach(entry => {
+        if (entry.isIntersecting) {
+          setActiveHeading(entry.target.id)
+        }
+      })
+    }
+
+    const observer = new IntersectionObserver(observerCallback, {
+      rootMargin: "0px 0px -80% 0px",
+      threshold: 0,
+    })
+
+    wysiwygContainers.forEach(container => {
+      const h2Elements = container.querySelectorAll("h2")
+      h2Elements.forEach(heading => observer.observe(heading))
+    })
+
+    const handleScroll = () => {
+      if (window.scrollY === 0) {
+        setActiveHeading("")
+      }
+    }
+
+    window.addEventListener("scroll", handleScroll)
```

Comment:
You could use [this library](https://usehooks-ts.com/react-hook/use-event-listener) which will unmount it for you.

Reply:
You might even want to debounce the callback or the value:
https://usehooks-ts.com/react-hook/use-debounce-value
https://usehooks-ts.com/react-hook/use-debounce-callback